File: main/041-20200127.go
package main

func firstMissingPositive(nums []int) int {
	foundZero := false
	for i := 0; i < len(nums); i++ {
		nums[i]--
		if nums[i] == 0 {
			foundZero = true
		}
	}

	if foundZero {
		return firstMissingPositive(nums) + 1
	}

	return 1
}

// func main() {
// 	fmt.Println(firstMissingPositive([]int{1, 2, 0}))
// 	fmt.Println(firstMissingPositive([]int{3, 4, -1, 1}))
// 	fmt.Println(firstMissingPositive([]int{7, 8, 9, 11, 12}))
// }

File: main/041-20200127.rb
# @param {Integer[]} nums
# @return {Integer}
def first_missing_positive(nums)
  return 1 if nums.empty?
  h = nums.each_with_object({}) { |n, h| h[n] = true }
  max = h.keys.max
  (1..max).each do |n|
    unless h[n]
      return n
    end
  end
  [max+1, 1].max
end

puts first_missing_positive([3,4,-1,1])
puts first_missing_positive([3,4,-1,1])
puts first_missing_positive([1,2,0])
puts first_missing_positive([7,8,9,11,12])
puts first_missing_positive([0])
puts first_missing_positive([-5])